proposals/chainer-operator-proposal.md

```diff
@@ -0,0 +1,222 @@
+## Motivation
+
+[Chainer][Chainer] is a Python-based, standalone open source framework for deep learning models. Chainer provides a flexible, intuitive, and high-performance means of implementing a full range of deep learning models, including state-of-the-art models such as recurrent neural networks and variational autoencoders.  
+
+[ChainerMN][ChainerMN] is an additional package for [Chainer][Chainer], which enables multi-node distributed deep learning in a scalable, flexible and easy way.  [ChainerMN][ChainerMN] currently supports MPI to initialize process groups or do collective communications(e.g. broadcast, all-reduce, etc.) among processes attending the distributed learning.  They are now planning to extend the support to other communication backends (e.g. [gloo][gloo] or other custom ones).
+
+Moreover, [Chainer][Chainer]/[ChainerMN][ChainerMN] achieved to [train ResNet-50 on ImageNet in 15 Minutes](https://arxiv.org/pdf/1711.04325.pdf) in the environment equipped with GPUs and InfiniBand FDR.  [The recent research](https://chainer.org/general/2018/05/25/chainermn-v1-3.html) revealed that [ChainerMN][ChainerMN]'s latest feature (Double-buffering and All-Reduce in half-precision float values) enables users to expect _almost_ linear scalability without sacrificing model accuracy even in environments (e.g. AWS) which doesn't equip InfiniBand.
+
+However, [Chainer][Chainer]/[ChainerMN][ChainerMN] currently does not have an operator/controller for Kubernetes. This proposal is aimed at defining what the operator should behave, and add it to Kubeflow.
+
+## Goals
+A Kubeflow user should be able to run training using [Chainer][Chainer]/[ChainerMN][ChainerMN] as easily as then can using Tensorflow/PyTorch.  This proposal is centered around a Kubernetes operator for [Chainer]/[ChainerMN]. A user should be able to run both single node with [Chainer][Chainer] and distributed training jobs with [ChainerMN][ChainerMN].
+
+This proposal defines the following:
+- A Chainer operator
+- A way to deploy the operator with ksonnet
+- A single pod Chainer example
+- A distributed (multiple pods) Chainer example
+
+## Non-Goals
+Currently, for the scope of this proposal, we won't be addressing the method for serving the model.
+
+## API (CRD and resulting objects)
+
+### Custom Resource Definition
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: ChainerJob
+metadata:
+  name: example-chainermn-job
+spec:
+  # only "mpi" is supported in the first scope
+  # "gloo", or custom backend will be supported in the future.
+  backend: mpi
+  replicaSpecs:
+  - replicas: 1
```

Comment:
Could we use map instead of slice as tf-operator does? Ref https://github.com/kubeflow/community/blob/master/proposals/tf-operator-design-v1alpha2.md

It will ease the implementation of the operator.

Reply:
done in https://github.com/kubeflow/community/pull/145/commits/9dd0bb3b2f39c02d550e3ec013b8a32b52212e84